### PR TITLE
Factor out loss_fn to share code with pipeline par

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #161
* #172
* __->__ #203
* #202
* #201

PP requires feeding a loss_fn into the schedule's step so that loss can
be computed per microbatch as part of the forward/backward scheduling.

As such, it is nice to define loss once and use it both in the non-pp
code that manually calls f/loss/b and also use it in the pp step().

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -149,6 +149,15 @@ def main(job_config: JobConfig):
         dp_rank,
     )
 
+    # loss fn can be shared by pipeline-parallel or non-pp execution
+    def loss_fn(pred, labels):
+        with (
+            loss_parallel()
```
Comment (wconstab) [inline train.py]: realized the .backward call needs to be under loss_parallel.  So i wonder if we can move loss_parallel to wrap the whole train loop or at least the parts that run forward+loss+backward.

this would make it easier to unify with PP where F/B are overlapped in different ways depending on schedule.

File: train.py [hunk 2]
```diff
@@ -149,6 +149,17 @@ def main(job_config: JobConfig):
...
 
+    # loss_parallel enables dispatching to efficient loss operators
+    loss_parallel_ctx = (
+        loss_parallel()
+        if parallel_dims.loss_parallel_enabled
+        else contextlib.nullcontext()
+    )
+
+    # loss fn can be shared by pipeline-parallel or non-pp execution
+    def loss_fn(pred, labels):
```
Comment (wanchaol) [inline train.py:160]: is this `loss_fn` have any actual difference compare to the `F.cross_entropy` be directly used by PP?

Comment (wconstab) [reply train.py:160]: views
